src/server/ua_server.c

```diff
@@ -544,65 +678,174 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
                 continue;
-            if(!UA_ByteString_equal(oldCertificate,
-                                    &session->channel->securityPolicy->localCertificate))
+
+            retval = sp->updateCertificateAndPrivateKey(sp, certInfo.certificate,
+                                                        certInfo.privateKey);
+            if(retval != UA_STATUSCODE_GOOD)
+                goto cleanup;
+        }
+        for(size_t j = 0; j < server->config.endpointsSize; j++) {
+            UA_EndpointDescription *ed = &server->config.endpoints[j];
+            UA_SecurityPolicy *sp =
+                getSecurityPolicyByUri(server, &server->config.endpoints[j].securityPolicyUri);
+            if(!sp) {
+                retval = UA_STATUSCODE_BADINTERNALERROR;
+                goto cleanup;
+            }
+
+            if(!UA_NodeId_equal(&certGroupId, &sp->certificateGroupId) ||
+               !UA_NodeId_equal(&certTypeId, &sp->certificateTypeId))
                 continue;
 
-            UA_LOCK(&server->serviceMutex);
-            UA_Server_removeSessionByToken(server, &session->authenticationToken,
-                                           UA_SHUTDOWNREASON_CLOSE);
-            UA_UNLOCK(&server->serviceMutex);
+            UA_String_clear(&ed->serverCertificate);
+            retval = UA_String_copy(&sp->localCertificate, &ed->serverCertificate);
+            if(retval != UA_STATUSCODE_GOOD)
+                goto cleanup;
         }
     }
 
-    /* Gracefully close all SecureChannels. And restart the
-     * BinaryProtocolManager once it has fully stopped. */
-    if(closeSecureChannels) {
-        UA_ServerComponent *binaryProtocolManager =
-            getServerComponentByName(server, UA_STRING("binary"));
-        if(binaryProtocolManager) {
-            binaryProtocolManager->notifyState = notifySecureChannelsStopped;
-            binaryProtocolManager->stop(server, binaryProtocolManager);
-        }
+    /* Add to the delayed callback list. Will be cleaned up in the next iteration. */
+    UA_DelayedCallback *dc = &transaction->dc;
+    dc->callback = sercureChannel_delayedClose;
+    dc->application = NULL;
+    dc->context = server;
+
+    UA_EventLoop *el = server->config.eventLoop;
+    el->addDelayedCallback(el, dc);
+
+    return UA_STATUSCODE_GOOD;
+
+cleanup:
+    UA_GDSTransaction_clear(transaction);
+    transaction = NULL;
+
+    return retval;
+}
+
+UA_StatusCode
+UA_Server_updateCertificate(UA_Server *server,
+                            const UA_NodeId certificateGroupId,
+                            const UA_NodeId certificateTypeId,
+                            const UA_ByteString *certificate,
+                            const UA_ByteString *issuerCertificates,
+                            const size_t issuerCertificatesSize,
+                            const UA_ByteString *privateKey,
+                            const UA_String *privateKeyFormat) {
```

Comment:
In case no key is provided, we have to find a matching key here by either:

- checking against the latest not-activated one (created by a CSR)
- checking against the current activated one (that is already used in the current application instance certificate)

If neither of them matches -> an error is expected.

An error must also be return, if the provided certificate and key do not belong to each other. This checks also seems to be missing.